File: resources/bin/bit_torrent_producer.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'time'
require 'kafka'

FILENAME = '/usr/lib/redborder/producers/json/bit_torrent.json'

# Initialize Kafka producer
kafka = Kafka.new(['localhost:9092'])
producer = kafka.producer
NAMESPACE_UUID = '352369f8-60fb-4b72-a603-d1d8393cca0a'
def modify_timestamp(json_data, real_last_time)
  puts json_data
  json_data['timestamp'] = json_data['timestamp'] + real_last_time
  json_data
end

def read_json_file(json_file)
  File.readlines(json_file).map { |line| JSON.parse(line) }
end

def replay_messages(json_file, producer)
  messages = read_json_file(json_file)
  real_last_time = messages.last['timestamp']

  messages.each do |message|
    # sleep_time = 
    # sleep(sleep_time) if sleep_time.positive?

    # Modify the timestamp and add to batch
    modified_message = modify_timestamp(message, real_last_time)

    begin
      producer.produce(modified_message.to_json, topic: 'rb_event_post_' + NAMESPACE_UUID)
      producer.deliver_messages
    rescue => e
      puts "Error delivering message: #{e.message}"
      sleep 1 # Add delay before retrying
      retry
    end    
    puts "Delivered message"
  end

  # Close the producer
  producer.shutdown
end

# Start the message replay process
replay_messages(FILENAME, producer)

File: resources/bin/bit_torrent_producer.py
#!/usr/bin/env python3

import json
import time
from kafka import KafkaProducer

FILENAME = "/usr/lib/redborder/producers/json/bit_torrent.json"
KAFKA_BROKER = "localhost:9092"
TOPIC_PREFIX = "rb_event_post_"
NAMESPACE_UUID = "352369f8-60fb-4b72-a603-d1d8393cca0a"

def read_json_file(json_file):
    with open(json_file, "r") as file:
        return [json.loads(line) for line in file]

def replay_messages(json_file, producer):
    messages = read_json_file(json_file)
    
    pre_timestamp = 0
    for message in messages:
        timestamp = message["timestamp"]
        message["timestamp"] = int(time.time())
            
        try:
            producer.send(TOPIC_PREFIX + NAMESPACE_UUID, json.dumps(message).encode("utf-8"))
            producer.flush()
        except Exception as e:
            print(f"Error delivering message: {e}")
            time.sleep(1)
            continue
        print("Delivered message")
        if pre_timestamp > 0:
            next_time = timestamp - pre_timestamp
            print(next_time)
            time.sleep(next_time)
        pre_timestamp = timestamp
    producer.close()

if __name__ == "__main__":
    producer = KafkaProducer(bootstrap_servers=[KAFKA_BROKER])
    replay_messages(FILENAME, producer)
